### ISSUE TITLE
Question: why do selected words flicker when setStyle

### ISSUE THREAD

**Issue by jzlhll** [FXMisc/RichTextFX#1055]
Hi, 

just like RichTextDemo.java, 
When select a color in "Text background" picker, the selection words will flicker a little.
It seems like padding or inner margin changed.

Please try it. 

You can paste with a code file and select words, choose the color picker 'Text background'.

What I expect is not has this little flicker.

the lastest version0.10.7 and java17.

**Comment by Col-E** [comment]
https://github.com/user-attachments/assets/8e265e90-e1e4-44a3-b39b-3efdd2bea881

Can still reproduce this on latest.

And on `RichTextDemo` here's a gif shortened to illustrate the slight jitter. See selected text, plus text to the right for slight movements. Left text before background color change is unmoved:

<img width="580" height="196" alt="Image" src="https://github.com/user-attachments/assets/5c821408-ab12-4ea4-bb48-c54611cd34bb" />

**Comment by Symeon94** [comment]
Hello,

I don’t know much about that issue, but I tried to reproduce the same behavior and I couldn’t get that slight movement either (neither in JavaKeywordsDemo nor in RichTextDemo).

Could you share some details about how you captured the video? That might help me understand the exact steps and see if I can reproduce it on my side.

I’m testing on Windows 10. Which OS are you using? I suspect the style behavior might vary depending on the platform.

Thanks for the info,
Syméon

**Comment by Col-E** [comment]
> I’m testing on Windows 10. Which OS are you using?

Windows 10 (22H2) on my system as well. Got a friend on Arch to also reproduce the same jitter effect.

> Could you share some details about how you captured the video? 

First video is from my own project Recaf. Second video is the demo and the change is _very_ small. Both illustrate the same behavior where specifying `-rtfx-background-color` causes subpixel movement.  Its so small that it is almost like a anti-alias change.

**Comment by Symeon94** [comment]
But how do you actually capture that.
So I assume you changed the original example by adding -rtfx-background-color? (could you share that css?)
I will see if I can replicate on my side.

**Comment by Symeon94** [comment]
Allow me to clarify my question, as I don't think I was clear enough. How did you practically record the video? I tried recording with my GPU software and stepping through the recording frame by frame, but I couldn't observe anything like what you described.

I also tried both setting the background color and changing the text fill color (with the following modification in `java-keywords.css`):

```css
.styled-text-area .text {
    -rtfx-background-color: #FFFFFF;
}

.styled-text-area .selection {
    -fx-fill: #557755;
}
```

However, I still see the same behavior and cannot reproduce the flicker you mentioned.

Let me know if you see anything relevant on your side that might explain this (version of Java, JavaFX, anything else).

**Comment by Col-E** [comment]
> But how do you actually capture that.

[ShareX](https://github.com/sharex/sharex) - which for encoding will use FFMPEG [under the hood](https://github.com/ShareX/ShareX/tree/0df9ca4d83eed9d2489048c539d7d1fc2860fdec/ShareX.ScreenCaptureLib/ScreenRecording).

It doesn't need to be a video, here's some images of before/after:

<img width="602" height="432" alt="Image" src="https://github.com/user-attachments/assets/1883f352-4b4a-4430-ab2f-6e763a169218" />

<img width="602" height="432" alt="Image" src="https://github.com/user-attachments/assets/469e0e1c-d85b-43d2-b53b-ec2b7359cc4f" />

And if you apply a difference algorithm between the images, the change is clear. You can see the text has moved slightly and so has the unstyled text to the right of it:

<img width="602" height="432" alt="Image" src="https://github.com/user-attachments/assets/d44abe45-0df8-4e5c-b008-348c99ebb6f3" />

> Java keywords demo

This does not dynamically update the background like the original issue steps describe so no observable change would occur here. Changing the text fill also does not appear to reproduce the issue.

> Let me know if you see anything relevant on your side that might explain this (version of Java, JavaFX, anything else).

- Java 22
- JavaFX 23.0.2
- GPU - NVIDIA RTX 4080 _(Default settings for system-wide anti-aliasing, any combination of changes does not affect observed behavior)_

**Comment by Symeon94** [comment]
Thanks for the tip about the PNG diff, I can see it now. I'll take a closer look.

My first observation is that when switching from one color (red for instance) to another (even to a dark gray), there's no visible change. However, when switching from a color back to the default black, the text shifts slightly.

My initial thought is that applying a color adds a property that has a subtle effect on the text layout, and that property is removed when switching back to the default color.

**Comment by Col-E** [comment]
> My initial thought is that applying a color adds a property that has a subtle effect on the text layout, and that property is removed when switching back to the default color.

I tried looking at the scene graph's styles via https://github.com/JonathanGiles/scenic-view and couldn't find anything super obvious with listed properties for the affected highlighted `TextExt`. Do note if you use this it expects `javafx-swing` on the runtime classpath.

**Comment by Col-E** [comment]
Also curiously, this does _not_ affect selection text which also sets a background color.

**Comment by Symeon94** [comment]
Looking at the code of the example, changing the color will apply a bunch of style dynamically on a StyleSpan. I would suspect that either it's the introduction of a substyle which creates this slight move (which explains why changing to another color doesn't do it) or that one of the default option added adds the space. I will try to add the default style on the whole text to see if it changes something.

**Comment by Symeon94** [comment]
OK, it doesn't seem to be that. If you zoom in very closely, you'll see that the letters themselves are in exactly the same position. The only difference is that the antialiasing changes ever so slightly. There must be some rounding difference happening somewhere that's causing JavaFX to produce a slightly different antialiasing result.

<img width="1861" height="359" alt="Image" src="https://github.com/user-attachments/assets/e0a7a448-612f-4573-9043-03929e65d152" />

Obviously that doesn't tell us why is there such difference. It could even be that it's the change of color that impacts the AA of JavaFX (so far I have no conclusion on that point).

**Comment by Symeon94** [comment]
Ok, I did the following experiment. I divided a text chunk of varying size (size 1,2,3,...) and stored them as `Text` in a `TextFlow`. If the AA varies based on the size of the `Text` we should see AA differences (as I think that's what is used in the underlying rendering for RichTextFX).

And I'm afraid the AA changes based on the size of the `Text`, meaning that the rendering issue is a JavaFX one.

Here is a picture of the same text of size 1/2/3, see how the rendering is clearly different for the end of "not" and the last "be":
<img width="803" height="683" alt="Image" src="https://github.com/user-attachments/assets/b0a2a74a-412a-46bf-812b-aabd9c57b27d" />
Again, this picture is pure JavaFX.

The [ugly] code if you want to reproduce:
```Java
        String text = "To be or not to be";
        for(int size = 1 ; size < text.length() ; size++) {
            TextFlow flow = new TextFlow();
            int count = text.length()/size;
            for (int j = 0; j < count; j++) {
                flow.getChildren().add(new Text(text.substring(j*size, j*size+size)));
            }
            if(text.length() % size > 0) {
                flow.getChildren().add(new Text(text.substring(count * size)));
            }
            AnchorPane.setTopAnchor(flow, size * 30.0);
            mainPane.getChildren().add(flow);
        }
```

**Comment by Col-E** [comment]
Oof, well on the plus side it means no library version bumps for RichTextFX😅

**Comment by Symeon94** [comment]
Well, it's good to always take the positive sides of every situation 😉

That said, I tried to look for possible solutions, but I don't think there is a real fix for this.

If I summarize the issue as I understand it:

JavaFX uses `double` values for text positioning, and those positions are taken into account when calculating anti-aliasing (AA). A tiny difference in the final glyph position can therefore change the rendered pixels, even if the text content is identical.

The only way to completely avoid this would be to guarantee that each piece of text is rendered at exactly the same position regardless of how the text is split.

For example, these two cases:

```
[To be or not to be]
[To be][ or not ][to be]
```

would need to produce exactly the same position for the middle section.

In other words, the text:

```
[ or not ]
```

would have to start at the exact same subpixel position as it did when it was part of the original single Text node.

Since splitting the text changes the layout process and creates separate text runs, JavaFX can produce slightly different positions and therefore slightly different AA.

So the problem is not only about pixel alignment; it is about preserving the exact glyph positioning of a continuous text run after it has been divided into multiple styled runs.

--------------------

By curiosity I had a look at how this issue is handled in other text rendering UI libraries. It seems that these render the glyphs first and then they proceed with rasterization. So whether you render in a way or another the position never varies.

The solution would be JavaFX to provide a way to style a text as a whole.

From my side, I consider that this issue is closed (at least from RichTextFX point of view).

**Comment by Col-E** [comment]
Agreed, with the core JavaFX reproduction sample its no longer a library issue.
